### PR TITLE
NXP Sabre_iMX6SX_1GB board support added.

### PR DESCRIPTION
**Pull request by mvarecha**

### CODE REVIEW

File: Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/Lcdif.c [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    status = ImxSetLcdIfClockRate(IMX6SX_PHYSADDR_LCDIF2, Timing->PixelClock);
+    if (EFI_ERROR(status)) {
+        DebugPrint(DEBUG_ERROR, "ConfigureLCD - ImxClkPwrLcdIfEnable failed");
+        return status;
+    }
+
+    sResetLCDIF();
+
+    // Debug code to allocate as much bandwidth as possible to LCDIF
+#if 0
```
Comment (hbeberman) [inline Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/Lcdif.c]: No `#if 0` code. Wrap it with an LCDIF_DEBUG define check.

File: Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/Lcdif.c [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    value32 |= IMX6SX_LCDIF_VDCTRL4_SYNC_SIGNALS_ON;
+    MmioWrite32(IMX6SX_PHYSADDR_LCDIF2_VDCTRL4, value32);
+
+    // clear FIFO clear
+    MmioWrite32(IMX6SX_PHYSADDR_LCDIF2_CTRL1_CLR, IMX6SX_LCDIF_CTRL1_FIFO_CLEAR);
+
+    // set run bit
+    MmioWrite32(IMX6SX_PHYSADDR_LCDIF2_RL_SET, IMX6SX_LCDIF_RL_RUN);
+
+#if 0
```
Comment (hbeberman) [inline Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/Lcdif.c]: No `#if 0` code. Wrap it with an LCDIF_DEBUG define check.

**Review by hbeberman — CHANGES_REQUESTED** [review]

File: Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/GraphicsOutputDxe.c [hunk 3]
```diff
@@ -0,0 +1,415 @@
+////
```
Comment (christopherco) [inline Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/GraphicsOutputDxe.c]: Please fix this header to follow the EDK2 copyright/license style. Applies to all files in this folder.

File: Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/GraphicsOutputDxe.c [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+
+//
+// Query for preferred timing
+//
+VOID
+GetPreferredTiming(
+    OUT IMX_DISPLAY_TIMING* PreferredTiming
+    )
+{
+//    EFI_STATUS status;
```
Comment (christopherco) [inline Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/GraphicsOutputDxe.c]: Please remove all commented out code.

File: Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/GraphicsOutputDxe.c [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+Exit:
+    return status;
+}
+
+// -------------------------------------------------------------------------------------------------------
+
+// Almost Identical to iMXVideoDxe.c VidGopBlt
+static
+EFI_STATUS
+sGfxBlt(
```
Comment (christopherco) [inline Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/GraphicsOutputDxe.c]: We recently introduced library code into iMXDisplayLib to implement Blt, QueryMode, and SetMode (https://github.com/ms-iot/imx-edk2-platforms/blob/imx/Silicon/NXP/iMXPlatformPkg/Library/iMXDisplayLib). I recommend using that library to reduce code copy/bloat.

See https://github.com/ms-iot/imx-edk2-platforms/pull/30 and https://github.com/ms-iot/imx-edk2-platforms/blame/imx/Platform/Udoo/UdooNeo_iMX6SX_1GB/Drivers/GraphicsOutputDxe/GraphicsOutputDxe.c#L272-L274 for examples of how to use this lib.

File: Platform/NXP/Sabre_iMX6SX_1GB/Library/iMX6BoardLib/iMX6BoardInit.c [hunk 6]
```diff
@@ -0,0 +1,1130 @@
...
+{
+    MmioWrite32(Addr, MmioRead32(Addr) | Bits);
+}
+
+static void MmioClrBits32(UINT32 Addr, UINT32 Bits)
+{
+    MmioWrite32(Addr, MmioRead32(Addr) & ~Bits);
+}
+
+/*
```
Comment (christopherco) [inline Platform/NXP/Sabre_iMX6SX_1GB/Library/iMX6BoardLib/iMX6BoardInit.c]: No commented out code please.

File: Platform/NXP/Sabre_iMX6SX_1GB/Library/iMX6BoardLib/iMX6BoardInit.c [hunk 7]
```diff
@@ -0,0 +1,1130 @@
...
+        return RETURN_SUCCESS;
+    }
+    ImxClkPwrInit();
+    // initialize default UEFI debug port early so we can use its debug output
+    UartInit();
+    SerialPortInitialize();
+    SerialPortWrite((UINT8 *)SERIAL_DEBUG_PORT_INIT_MSG, (UINTN)sizeof (SERIAL_DEBUG_PORT_INIT_MSG));
+//  SerialPrint("\nSerial Port Init\n\r");
+    I2cInit();
+//  HdmiInit();
```
Comment (christopherco) [inline Platform/NXP/Sabre_iMX6SX_1GB/Library/iMX6BoardLib/iMX6BoardInit.c]: Remove commented out code

File: Platform/NXP/Sabre_iMX6SX_1GB/Library/iMX6BoardLib/iMX6BoardLib.inf [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  ArmLib
+  MemoryAllocationLib
+  SerialPortLib
+  iMX6ClkPwrLib
+  iMX6AudioLib
+
+[Sources.common]
+  $(BOARD_LIB_COMMON_DIR)/iMX6BoardHelper.S    | GCC
+  $(BOARD_LIB_COMMON_DIR)/iMX6BoardMem.c
+#  $(BOARD_LIB_COMMON_DIR)/iMX6Common.c
```
Comment (christopherco) [inline Platform/NXP/Sabre_iMX6SX_1GB/Library/iMX6BoardLib/iMX6BoardLib.inf]: Remove commented out code

**Review by christopherco — CHANGES_REQUESTED** [review]

File: Platform/NXP/Sabre_iMX6SX_1GB/Library/iMX6BoardLib/iMX6BoardInit.c [hunk 9]
```diff
@@ -0,0 +1,1130 @@
...
+  Initialize SDHC modules on the SOC and perform required pin-muxing
+**/
+VOID SdhcInit ()
+{
+    DebugPrint(0xFFFFFFFF, "VOID SdhcInit()\n");
+
+    //
+    // uSDHC2: SDIO Socket
+    //
+    ImxPadConfig(IMX_PAD_SD2_CLK, IMX_PAD_CFG_SD2_CLK_SD2_CLK);
```
Comment (christopherco) [inline Platform/NXP/Sabre_iMX6SX_1GB/Library/iMX6BoardLib/iMX6BoardInit.c:650]: Are these already Muxed in U-Boot? If so, please don't mux them again in EDK2.

Applies to all IP. We want to reduce the amount of code in firmware. Typically U-Boot enables specific IP like Sdhc and Enet and muxes the PADs during its init flow. If U-Boot already enabled a specific IP, don't duplicate the code in EDK2.

File: Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/Lcdif.c (hunk 1 above)
Comment (mvarecha) [reply Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/Lcdif.c]: Fixed.

File: Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/Lcdif.c (hunk 2 above)
Comment (mvarecha) [reply Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/Lcdif.c]: Fixed

File: Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/GraphicsOutputDxe.c (hunk 3 above)
Comment (mvarecha) [reply Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/GraphicsOutputDxe.c]: Fixed.

File: Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/GraphicsOutputDxe.c (hunk 4 above)
Comment (mvarecha) [reply Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/GraphicsOutputDxe.c]: Fixed.

File: Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/GraphicsOutputDxe.c (hunk 5 above)
Comment (mvarecha) [reply Platform/NXP/Sabre_iMX6SX_1GB/Drivers/GraphicsOutputDxe/GraphicsOutputDxe.c]: Driver updated.

File: Platform/NXP/Sabre_iMX6SX_1GB/Library/iMX6BoardLib/iMX6BoardInit.c (hunk 6 above)
Comment (mvarecha) [reply Platform/NXP/Sabre_iMX6SX_1GB/Library/iMX6BoardLib/iMX6BoardInit.c]: Fixed.

File: Platform/NXP/Sabre_iMX6SX_1GB/Library/iMX6BoardLib/iMX6BoardInit.c (hunk 7 above)
Comment (mvarecha) [reply Platform/NXP/Sabre_iMX6SX_1GB/Library/iMX6BoardLib/iMX6BoardInit.c]: Fixed.

File: Platform/NXP/Sabre_iMX6SX_1GB/Library/iMX6BoardLib/iMX6BoardLib.inf (hunk 8 above)
Comment (mvarecha) [reply Platform/NXP/Sabre_iMX6SX_1GB/Library/iMX6BoardLib/iMX6BoardLib.inf]: Fixed.

**Review by hbeberman — APPROVED** [review]

**Comment by dmcilvaney** [comment]
If the board is ready, you can add a continuous integration target to https://github.com/ms-iot/imx-iotcore/blob/9e82401a927cdb396c43d892b8e74106f0ebb953/ci/azure-pipelines.yml#L137 .

**Review by christopherco — APPROVED** [review]

**Pull request merged**